### DISCUSSION TITLE
Is there a propagator that provides perfectly matched layer boundary condition

### DISCUSSION THREAD

**Discussion by tomsturges** [chromatix-team/chromatix#174] [Q&A] [answered]
Hi everyone. What I am looking for is a propagator that gives a numerical approximation of the propagation of a field from one plane to another with perfectly matched layers (i.e. any field that gets to the edge of the domain is perfectly absorbed, so no wrap around effects). This would allow one to propagate to another plane where lots of the field might be leaking out, so there's lots of field at the edges, but this transform is still a faithful approximation within the simulation domain. My understanding is that one can use something like the angular spectrum method but then embedding the linear convolution inside a circular convolution with double the array length to implement this exactly. However I couldn't quite figure out how to accomplish this with Chromatix. Could you help me out? Thanks :)

**Comment by GJBoth** [comment]
If you set the [`N_pad`](https://chromatix.readthedocs.io/en/latest/api/functional/#chromatix.functional.propagation.asm_propagate) equal to the incoming field's size you'll do a linear convolution - is that what you're looking for?

**Comment by tomsturges** [comment]
Hi @GJBoth, thanks for the response. This doesn't seem to behave as I expect, or as I'm looking for rather.

Here is some simple code to propagate a `gaussian_planewave`:

```python
import chromatix as cx
from chromatix.functional import gaussian_plane_wave, asm_propagate
import matplotlib.pyplot as plt
import numpy as np
import cmcrameri.cm as cmc

field = gaussian_plane_wave((500, 500), 0.1, 0.45, 14, kykx=(2, 0))
field = asm_propagate(field, z=np.linspace(0, 1000, 5), n=1, pad_width=0)
I = field.intensity

fig, ax = plt.subplots(1, 5)
for i, im in enumerate(I):
    ax[i].imshow(im, cmap=cmc.davos, origin='lower')
    ax[i].set_title(f'Frame {i}')
    ax[i].axis('off')
plt.tight_layout()
plt.show()
```

We see the wave-packet wrapping around the edges of the domain

<img width="628" height="362" alt="image" src="https://github.com/user-attachments/assets/6ecf7da6-2ccd-4176-a1d6-bcc5c4a03080" />

Ok so now I tried changing `N_pad=500` like you suggest, i.e. update the one line to:

`field = asm_propagate(field, z=np.linspace(0, 1000, 5), n=1, pad_width=500)`

But this just gives a larger simulation domain and still causes wrap around effects at the edges of the larger domain.

<img width="628" height="351" alt="image" src="https://github.com/user-attachments/assets/982d4e2a-03b6-4c0a-98fd-633575082b10" />

I am rather looking for something where the wavepacket "disappears" into the edge of the domain with perfect absorption.

**Reply by diptodip** [reply]
The propagation methods do have absorbing boundaries (Tukey and super-Gaussian). See the absorbing_boundary and absorbing_boundary_width arguments. Does that work for you?

The other thing you would want with padding is to set the mode to "same" to crop back to the original size to prevent wrapping.

**Comment by tomsturges** [comment]
Thanks @diptodip. However, neither of those things seems to absorb anything for me. Not sure what I'm doing wrong.

```python
import chromatix as cx
from chromatix.functional import gaussian_plane_wave, asm_propagate
import matplotlib.pyplot as plt
import numpy as np
import cmcrameri.cm as cmc

zlist = np.linspace(0, 1000, 10)
field = gaussian_plane_wave((500, 500), 0.1, 0.45, 14, kykx=(2, 0))
### PAY ATTENTION TO THE LINE BELOW ###
field = asm_propagate(field, z=zlist, n=1, pad_width=500, mode='same', absorbing_boundary='tukey', absorbing_boundary_width=0.65)
I = field.intensity

fig, ax = plt.subplots(2, 5)
ax = ax.ravel()
for i, im in enumerate(I):
    ax[i].imshow(im, cmap=cmc.davos, origin='lower')
    ax[i].set_title(f'Frame {i}')
    ax[i].axis('off')
plt.tight_layout()
plt.show()
```

<img width="1076" height="597" alt="image" src="https://github.com/user-attachments/assets/14a0a278-e59d-4415-abab-2914ebb4f776" />

**Reply by diptodip** [reply]
I think I've found the issue, I'll update soon! Thanks for testing.

**Reply by diptodip** [reply, marked as answer]
@tomsturges The fixes to the absorbing boundaries (and your latest JIT issue) are both up on `dd/dev` (will be merged to main soon, but you could also `$ pip install git+https://github.com/chromatix-team/chromatix.git@dd/dev` if you want to try before that).

But as for what you actually want to do here, I think the absorbing boundaries that we have may not be what you want. Just padding should actually take care of it. The problem is that you are calculating all the distances in parallel, and those distances (especially the farther ones) wrap even with the amount of padding you've chosen.

You have three options that would work out.

1. Use some padding, but propagate in small steps instead of to the farthest distances in one step:
```python
import chromatix as cx
from chromatix.functional import gaussian_plane_wave, asm_propagate
import matplotlib.pyplot as plt
import jax.numpy as jnp
import numpy as np

zlist = np.linspace(0, 1000, 10)
field = gaussian_plane_wave((500, 500), 0.1, 0.45, 14, kykx=(2, 0))
### PAY ATTENTION TO THE LINES BELOW ###
fields = [field.u]
for i in range(zlist.size - 1):
    field = asm_propagate(field, z=111.11111111, n=1, pad_width=500 // 2, mode='same')
    fields.append(field.u)
field = field.replace(u=jnp.stack(fields))
I = field.intensity

fig, ax = plt.subplots(2, 5)
ax = ax.ravel()
for i, im in enumerate(I):
    ax[i].imshow(im, cmap="cividis", origin='lower', vmin=I.min(), vmax=I.max())
    ax[i].set_title(f'Frame {i}')
    ax[i].axis('off')
plt.tight_layout()
plt.show()
```
(I notice some artifacting near the edge of this one.)
<img width="629" height="354" alt="image" src="https://github.com/user-attachments/assets/343b9f57-4f91-4e7d-aff2-87b3d9347783" />

2. Use a much larger amount of padding and still propagate all distances in parallel:
```python
import chromatix as cx
from chromatix.functional import gaussian_plane_wave, asm_propagate
import matplotlib.pyplot as plt
import jax.numpy as jnp
import numpy as np

zlist = np.linspace(0, 1000, 10)
field = gaussian_plane_wave((500, 500), 0.1, 0.45, 14, kykx=(2, 0))
### PAY ATTENTION TO THE LINE BELOW ###
field = asm_propagate(field, z=zlist, n=1, pad_width=1000, mode='same')
I = field.intensity

fig, ax = plt.subplots(2, 5)
ax = ax.ravel()
for i, im in enumerate(I):
    ax[i].imshow(im, cmap="cividis", origin='lower', vmin=I.min(), vmax=I.max())
    ax[i].set_title(f'Frame {i}')
    ax[i].axis('off')
plt.tight_layout()
plt.show()
```
(I think this one actually looks the best, but it's a lot more expensive.)
<img width="629" height="354" alt="image" src="https://github.com/user-attachments/assets/98568548-6b1b-4f8b-a538-0fcc4747bf42" />

3. Use a smaller input source, no padding, and apply an absorbing boundary:
```python
import chromatix as cx
from chromatix.functional import gaussian_plane_wave, asm_propagate
import matplotlib.pyplot as plt
import jax.numpy as jnp
import numpy as np

zlist = np.linspace(0, 1000, 10)
### SMALLER INPUT ###
field = gaussian_plane_wave((500, 500), 0.1, 0.45, 4, kykx=(2, 0))
### PAY ATTENTION TO THE LINES BELOW ###
fields = [field.u]
for i in range(zlist.size - 1):
    field = asm_propagate(field, z=111.11111111, n=1, pad_width=0, absorbing_boundary="tukey", absorbing_boundary_width=0.3)
    fields.append(field.u)
field = field.replace(u=jnp.stack(fields))
I = field.intensity

fig, ax = plt.subplots(2, 5)
ax = ax.ravel()
for i, im in enumerate(I):
    ax[i].imshow(im, cmap="cividis", origin='lower', vmin=I.min(), vmax=I.max())
    ax[i].set_title(f'Frame {i}')
    ax[i].axis('off')
plt.tight_layout()
plt.show()
```
(This is a very different result from your original input, because otherwise the boundaries will cut off the input too early.)
<img width="629" height="354" alt="image" src="https://github.com/user-attachments/assets/99aaf441-3282-4f3d-9195-3588b521deb5" />

**Reply by tomsturges** [reply]
Ok I finally understand my misconceptions. You're of course correct that wrap around happens eventually no matter what, unless you explicitly add some absorbing regions. Thanks

**Reply by diptodip** [reply]
Great! Let me know if we can help with anything else.
